### PR TITLE
lirc: redesign, fix

### PR DESCRIPTION
**Pull request by FernetMenta**
Register will land in OptionalsReg

This drops sending command from Kodi to Lirc. I don't think this makes much sense.

### CODE REVIEW

**Comment by lrusak** [comment]
Can we register lirc in linux optionals now?

**Comment by FernetMenta** [comment]
@MilhouseVH this version has lirc only registered for X11. Testing until OptionslReg for Linux is merged (other PR). So no surprise that it's not working for PI.
But this log for X11 looks very strange. Looks like something creates one CLirc object after another.

**Comment by FernetMenta** [comment]
> Can we register lirc in linux optionals now?

@lrusak the other PR need to go in first.

File: xbmc/input/InputManager.h
Comment (afedchin) [inline xbmc/input/InputManager.h]: this isn't needed I suppose.

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: fixed, thanks

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/platform/linux/input/LIRC.h
Comment (garbear) [inline xbmc/platform/linux/input/LIRC.h]: Can you move the IInput interface to a new file (maybe in an interfaces/ subdir) to avoid including this header here?

Comment (FernetMenta) [reply xbmc/platform/linux/input/LIRC.h]: I am not happy with this either and it was one of the last things I did for this PR. Lirc is now pretty self-managed and does not need InputManager. InputManager is not a good container for Lirc but I had no better idea. Suggestions welcome.

File: xbmc/platform/linux/input/LIRC.cpp
Comment (garbear) [inline xbmc/platform/linux/input/LIRC.cpp]: No introduction of globals in the input code

Comment (FernetMenta) [reply xbmc/platform/linux/input/LIRC.cpp]: Capp is not a real global, it is the container for all other objects. It is CApp that owns the main thread and messages from input or windowing needs to be sent to Capp.
Originally ApplicationMessenger was in charge to accept messages for the main thread but it has been abusued for a ot of other things.
Capp needs an inbound port for messages that solely serves this purpose.

Comment (garbear) [reply xbmc/platform/linux/input/LIRC.cpp]: Then lets add an interface for inbound messages instead of referencing g_application directly.

File: xbmc/platform/linux/input/LIRC.cpp
Comment (Rechi) [inline xbmc/platform/linux/input/LIRC.cpp]: unused variable

File: xbmc/platform/linux/input/LIRC.cpp
Comment (Rechi) [inline xbmc/platform/linux/input/LIRC.cpp]: `lirc_init(static_cast<char*>("kodi"), 0);`

Comment (garbear) [reply xbmc/platform/linux/input/LIRC.cpp]: you mean `const_cast` I think

Comment (Rechi) [reply xbmc/platform/linux/input/LIRC.cpp]: maybe, haven't tested what is actually need to get rid of the warning and that it still compiles

File: xbmc/platform/linux/input/LIRC.cpp
Comment (FernetMenta) [reply xbmc/platform/linux/input/LIRC.cpp]: If you want this is the scope of this PR, fixing Lirc will almost affect all platforms. g_application.OnEvent is used by windowing.

File: xbmc/input/InputManager.h
Comment (afedchin) [inline xbmc/input/InputManager.h]: I didn't get the idea why manager should hold pointer to an input. They are independent now. in my opinion a creator should hold a pointer and manage an input lifetime.

File: xbmc/platform/linux/input/LIRC.cpp
Comment (afedchin) [reply xbmc/platform/linux/input/LIRC.cpp]: or we can invert dependency. i.e. a creator depends on an input and the creator depends on CApp. CLirc produces events, the creator subscribe to receive these events, the creator forwards events from the input to the CApp. this IoC makes weak dependency between an input and CApp and only creator bind them

File: xbmc/input/InputManager.h
Comment (FernetMenta) [reply xbmc/input/InputManager.h]: I agree, see comments above: Lirc.h include....

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: btw. almost independent. manager sill holds the config for the mapping, i.e. Lircmap.xml

Comment (afedchin) [reply xbmc/input/InputManager.h]: no, they are absolutely independent. mappings may be different with the same implementation. I agree that using Lirc without mapping has no sense but in my view this is very weak reference.

Comment (afedchin) [reply xbmc/input/InputManager.h]: as an example Xbox remote control has no mappings at all

**Comment by Rechi** [comment]
`liblirc-dev` build dependency should be added to the docs.

File: xbmc/input/InputManager.h
Comment (garbear) [reply xbmc/input/InputManager.h]: If IInput depends on InputManager, then it's not safe to destruct it from IInputManager's destructor. It needs a lifetime outside IInputManager so it can be unregistered on shutdown.

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: Lirc can be destructed at any time

Comment (garbear) [reply xbmc/input/InputManager.h]: Not after InputManager, or `CServiceBroker::GetInputManager().TranslateLircRemoteString()` will segfault

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: InputManger needs to register with a pointer at ServiceBroker.

Comment (garbear) [reply xbmc/input/InputManager.h]: Good point, that would work. However, I've been wondering for a while, how do we prevent race conditions between when the pointer is accessed via `CServiceBroker::GetInputManager()` and when it is used? Return a shared pointer?

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: Yes, shared ptr shoul work

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: Yes, shared ptr shoul work

Comment (garbear) [reply xbmc/input/InputManager.h]: But shared_ptr destroys init/deinit order. If InputManager has a dangling reference, it could be destroyed after something that it depends on is destroyed. We had this same problem with CPeripherals before.

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: In final stage the order of init/deinit should not matter.

Comment (garbear) [reply xbmc/input/InputManager.h]: how do we get from here to there?

Comment (FernetMenta) [reply xbmc/input/InputManager.h]: See this pr. Now lirc is independent from order of init. AE is next, only needs a little change. Powermanager shoulndˋt be a big deal either. Why donˋt you fix peripherials or some other component?

**Comment by garbear** [comment]
@FernetMenta #470

**Review by garbear — APPROVED** [review]
Good to go after squashing my squashme's into your commit.

**Comment by FernetMenta** [comment]
now lirc is completely independent from InputManager

**Comment by FernetMenta** [comment]
@popcornmix @MilhouseVH could you test this on pi please?
@pkerling not sure where you would like to register lirc on Wayland

**Comment by lrusak** [comment]
@FernetMenta thanks for doing the gbm lirc registration 👍

File: xbmc/platform/linux/OptionalsReg.h
Comment (pkerling) [inline xbmc/platform/linux/OptionalsReg.h]: unused?

Comment (FernetMenta) [reply xbmc/platform/linux/OptionalsReg.h]: thanks, fixed

**Comment by pkerling** [comment]
@FernetMenta Do you see a reason for not doing it like the other windowing systems?

I personally don't get why windowing has to deal with lirc at all, for me it's more of a platform thing.

**Comment by FernetMenta** [comment]
> Do you see a reason for not doing it like the other windowing systems?

nope but Wayland has a lot more members. should I add it to the end?

> I personally don't get why windowing has to deal with lirc at all, for me it's more of a platform thing.

a headless system does not need a remote control like lirc.

**Comment by pkerling** [comment]
> nope but Wayland has a lot more members. should I add it to the end?

Don't think there's any similar members at the moment, so yes that would be alright.

> a headless system does not need a remote control like lirc.

Of course, but that does not make lirc the responsibility of the win system :-) Win systems in Kodi already deal with way too much stuff.

**Comment by MilhouseVH** [comment]
RPi is working now, thanks, but messages are split over two lines: http://ix.io/17Jq

For example:
```
09:19:47.849 T:1864364928   DEBUG: LIRC: - NEW 6c 0 KEY_DOWN devinput
                                             (KEY_DOWN)
09:19:47.854 T:1941860368   DEBUG: HandleKey: 167 (0xa7, obc88) pressed, action is Down
```
Is this intentional?

Not sure what the errors at the end are about, probably something to do with #13754 - will report it there.

X86_64 still has the LIRC restart issue: http://ix.io/17Ju

**Comment by MilhouseVH** [comment]
I restarted kodi (`systemctl stop kodi`, waited 5-10 seconds, `systemctl start kodi`) on x86_64 and it's stuck on the Kodi splash: http://ix.io/17Jx

`eventlircd` is stuck at 100% CPU:

image

Rebooting will allow Kodi to start successfully. This seems to be reproducible.

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/AppParamParser.cpp
Comment (FernetMenta) [inline xbmc/AppParamParser.cpp]: for those who used this: the correct way to change lirc's defailt socket path via this environment variable LIRC_SOCKET_PATH

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
build error (test suite) not related

**Comment by fritsch** [comment]
@gelotus why not PR this minimal piece of code needed?

**Pull request merged**
